File: examples/consensus/extract_api.py
# ---------------------------------------------
## Quick example: Extract structured data using Retab's all-in-one `.parse()` method.
# ---------------------------------------------

import os
import json

from dotenv import load_dotenv

from retab import Retab

# Load environment variables
load_dotenv()

retab_api_key = os.getenv("RETAB_API_KEY")
assert retab_api_key, "Missing RETAB_API_KEY"

# Retab Setup
reclient = Retab(api_key=retab_api_key)

# Document Extraction via Retab API
response = reclient.documents.extract(
    document="../../assets/code/invoice.jpeg",
    model="gpt-4.1",
    json_schema="../../assets/code/invoice_schema.json",
    modality="native",
    image_resolution_dpi=96,
    browser_canvas="A4",
    temperature=0.0,
    n_consensus=4,  # Number of parallel extraction runs for consensus voting. Values > 1 enable consensus mode.
)

# Output
print("CONSENSUS EXTRACTION RESULTS")

for i in range(len(response.choices)):
    print(f"\nConsensus Result #{i + 1}:")
    print("-" * 40)
    try:
        content = response.choices[i].message.content
        if content and (content.strip().startswith("{") or content.strip().startswith("[")):
            parsed = json.loads(content)
            print(json.dumps(parsed, indent=2, ensure_ascii=False))
        else:
            print(content if content else "No content")
    except (json.JSONDecodeError, AttributeError):
        content = response.choices[i].message.content
        print(content if content else "No content")

# Display likelihoods with better formatting
if hasattr(response, "likelihoods") and response.likelihoods:
    print("\nCONSENSUS LIKELIHOODS:")
    print("-" * 40)

    # Handle both list and dict formats for likelihoods
    if isinstance(response.likelihoods, list):
        for i, likelihood in enumerate(response.likelihoods):
            print(f"Result #{i + 1}: {likelihood:.4f}")

        # Show the most confident result
        if response.likelihoods:
            best_idx = response.likelihoods.index(max(response.likelihoods))
            print(f"\n🎯 Most confident result: #{best_idx + 1} (likelihood: {max(response.likelihoods):.4f})")
    else:
        # Handle dict format
        for key, likelihood in response.likelihoods.items():
            print(f"{key}: {likelihood:.4f}")

File: examples/node/consensus/extract_api.js
// ---------------------------------------------
// Quick example: Extract structured data using Retab's all-in-one `.parse()` method.
// ---------------------------------------------

import { Retab } from '@retab/node';
import { config } from 'dotenv';

// Load environment variables
config();

const retabApiKey = process.env.RETAB_API_KEY;
if (!retabApiKey) {
  throw new Error('Missing RETAB_API_KEY');
}

// Retab Setup
const reclient = new Retab({ api_key: retabApiKey });

// Document Extraction via Retab API
const response = await reclient.documents.extract({
  document: '../../assets/code/invoice.jpeg',
  model: 'gpt-4.1',
  json_schema: '../../assets/code/invoice_schema.json',
  modality: 'native',
  image_resolution_dpi: 96,
  browser_canvas: 'A4',
  temperature: 0.1,
  n_consensus: 4, // Number of parallel extraction runs for consensus voting. Values > 1 enable consensus mode.
});

// Output
console.log('CONSENSUS EXTRACTION RESULTS');

for (let i = 0; i < response.choices.length; i++) {
  console.log(`\nConsensus Result #${i + 1}:`);
  console.log('-'.repeat(40));
  try {
    const content = response.choices[i].message.content;
    if (content && (content.trim().startsWith('{') || content.trim().startsWith('['))) {
      const parsed = JSON.parse(content);
      console.log(JSON.stringify(parsed, null, 2));
    } else {
      console.log(content || 'No content');
    }
  } catch (error) {
    const content = response.choices[i].message.content;
    console.log(content || 'No content');
  }
}

// Display likelihoods with better formatting
if (response.likelihoods) {
  console.log('\nCONSENSUS LIKELIHOODS:');
  console.log('-'.repeat(40));

  // Handle both array and object formats for likelihoods
  if (Array.isArray(response.likelihoods)) {
    response.likelihoods.forEach((likelihood, i) => {
      console.log(`Result #${i + 1}: ${likelihood.toFixed(4)}`);
    });

    // Show the most confident result
    if (response.likelihoods.length > 0) {
      const maxLikelihood = Math.max(...response.likelihoods);
      const bestIdx = response.likelihoods.indexOf(maxLikelihood);
      console.log(`\n🎯 Most confident result: #${bestIdx + 1} (likelihood: ${maxLikelihood.toFixed(4)})`);
    }
  } else {
    // Handle object format
    for (const [key, likelihood] of Object.entries(response.likelihoods)) {
      const formattedLikelihood = typeof likelihood === 'number' ? likelihood.toFixed(4) : likelihood;
      console.log(`${key}: ${formattedLikelihood}`);
    }
  }
}